### ISSUE TITLE
Safe::parse_and_resolve_url() seems wrong.  should be recursive and merge paths + query string

### ISSUE THREAD

**Issue by dan-da** [maidsafe/safe-api#514]
It is possible to link an nrs name to another nrs name.  This means that one can have a chain of links many links deep.    However, Safe::parse_and_resolve_url() only resolves the very first link.  As such, the caller will receive only a partially resolved XorUrlEncoder instance.

Here is a demonstration of creating a chain with 3 links:

```
$ safe files put --recursive ../testdata
FilesContainer created at: "safe://hnyynywwzjn6jrznz1iqqimd6fk5moqahmfysy1b8sqjrb6r3oi6ud13kabnc"

$ safe nrs create --link safe://hnyynywwzjn6jrznz1iqqimd6fk5moqahmfysy1b8sqjrb6r3oi6ud13kabnc?v=0 testdata
New NRS Map for "safe://testdata" created at: "safe://hnyydyzi7rz86yhfmfytfyw34ryxdt1b1j8keue79d183qgf44a5j6sc3ebqh"

$ safe nrs create --link safe://testdata/subfolder?v=0 subfolderNew NRS Map for "safe://subfolder" created at: "safe://hnyydyiefkkfpnsfbuigp1pom3p1rxgyobwaqecz1y5q497myqyowq6wacbqh"
+  subfolder  safe://testdata/subfolder?v=0

$ safe nrs create --link safe://subfolder/sub2.md?v=0 sub2.md
New NRS Map for "safe://sub2.md" created at: "safe://hnyydywwd93573xw9xa4zo7fptrciitdzhrfeehx6ua75uyr5isku6j891bqh"
+  sub2.md  safe://subfolder/sub2.md?v=0
```

Now let's try the first link, which works:

```
$ safe files ls safe://testdata
Files of FilesContainer (version 0) at "safe://testdata":
Files: 0   Size: 0   Total Files: 5   Total Size: 45
SIZE  CREATED               MODIFIED              NAME 
45    2020-04-11T16:44:44Z  2020-04-11T16:44:44Z  testdata/
```

But when we try the second deep link, it already fails:

```
$ safe files ls safe://subfolder
[2020-04-11T16:47:29Z ERROR safe] safe-cli error: Make sure the URL targets a FilesContainer.
[Error] InvalidXorUrl - Failed to decode XOR-URL: UnkownBase
```

and of course further links fail:

```
$ safe files ls safe://sub2.md
[2020-04-11T16:47:54Z ERROR safe] safe-cli error: Make sure the URL targets a FilesContainer.
[Error] InvalidXorUrl - Failed to decode XOR-URL: UnkownBase
```

I propose that the parse_and_resolve_url() should accept a "max_links" param and return a fully resolved XorUrlEncoder with the all the paths and query strings merged together.

eg, consider:

```
safe://testdata?v=1    -->  safe://xor.../testdata?v=3
safe://subfolder?v=6  --> safe://testdata/subfolder?v=1&arg=val
safe://sub2.md?v=0   --> safe://subfolder/sub2.md?v=6&key=val&arg=other
```

Now, if a caller asks to resolve safe://sub2.md?foo=bar, the fully resolved URL should be:

`safe://xor.../testdata/subfolder/sub2.md?key=val&arg=other&foo=bar`

So for nrs-url and linked-url, the resolution rules are:
1. The path of a nrs-url is appended to the path of linked-url
2. The v param of nrs-url is used to lookup linked-url, then v from linked-url is preserved.
3. Any other query string param in nrs-url overwrites same param in linked-url.
4. A fragment identifier in nrs-url replaces a fragment identifier in linked-url, if present.

**Comment by bochaco** [comment]
I tihnk it sounds good, we have to do a similar enhancement to the dog command, internally we have all the logic to do the recursive resolution but we are not returning the complete chain form any API or command like the dog. The dog command was thought to show the complete chain of resolution, see #518 for related discussion.

So I think you should be able to internally use https://github.com/maidsafe/safe-api/blob/master/safe-api/src/api/app/fetch.rs#L147, you may need to add support for the paths resolution, and then just return the complete list rather than one element. Note this implementation already solves the issue of using async, i.e. it's not recursive but iterative so it works with async.

> parse_and_resolve_url() should accept a "max_links" param

I think it should be enough to simply have a hard limit, like what we have in that function I mentioned above: https://github.com/maidsafe/safe-api/blob/master/safe-api/src/api/app/fetch.rs#L169, we could allow an optional arg to override the value but I fear that could be used by an app to break someone's browser tab or some type of attack, in which case you'd need a hard max even if the user can provide a value, so perhaps just one hard max hard-coded is all we need.

> 3. Any other query string param in nrs-url overwrites same param in linked-url.
> 4. A fragment identifier in nrs-url replaces a fragment identifier in linked-url, if present.

I think we shouldn't overwrite anything on the links we obtain from the NRS containers/resolution steps, as that would break perm web, or at least it breaks it's immutability perhaps, if a query string or fragment is the same perhaps give priority to the one found in the link obtained from NRS?

**Comment by bochaco** [comment]
I'm seeing we do have some path resolution there already, so it may be just fine even with paths: https://github.com/maidsafe/safe-api/blob/master/safe-api/src/api/app/fetch.rs#L225.

**Comment by dan-da** [comment]
I'm seeing we do have some path resolution there already, so it may be just fine even with paths: https://github.com/maidsafe/safe-api/blob/master/safe-api/src/api/app/fetch.rs#L225.

**Comment by bochaco** [comment]
> why do we need &foo=bar on a FileContainer ??

I just think it'll be breaking the immutability of the URL, the same way we don't want anyone to override the version or path the link contains, we shouldn't allow to override anything else? I would simply ignore any query arg/fragment that collides with what is in the link. If someone decides to put a query arg in the NRS link it shouldn't be possible to change it, in fact the version is a query arg, so you'd be making exceptions already

**Comment by dan-da** [comment]
But we are overriding the path, in the sense that we are appending to it, to create a fully resolved path.

**Comment by bochaco** [comment]
Yes but all that is being done still within immutable URLs, when you append it to that other path, that version of the URL will always be append it with the same path, you cannot override that version of the URL in any form, that's what I mean. If you can override the same version with some path or query string, or fragment I see like it's breaking the immutability of that version of that URL.

**Comment by dan-da** [comment]
okay, let me see if I can summarize/illustrate your view/rules here:

1. The path of a nrs-url is appended to the path of linked-url
2. ignore query string param in nrs-url that conflicts with same name param in linked-url
3. append/use query string param in nrs-url that does not conflict with param in linked-url
4. fragments are irrelevant (user-agent only) and should not be stored.  (added by me)

I suppose that is a viable ruleset.   It does not require a special case for ?v=<x>.

I find it less composeable than the ruleset I proposed.  Where each higher-level link, up to and including the user typing into browser urlbar, can override.  But it does have persistence going for it.

I think we need to be careful and thoughtful here, maybe get more input from team and community because it is one of those simple/subtle things where we probably aren't seeing all use-cases and ramifications that will occur as more and more gets built on top, either way.

Perhaps one issue for me is that I'm unclear how query strings will be used beyond ?v.  Given that all the data is basically static, so not handling dynamic requests like a normal webserver.  In a dynamic request model, it's important that user/app can modify query params.   But perhaps not in this system......?

**Comment by bochaco** [comment]
Yes, we need to get more feedback from others.

We need to decide if even that you'd be retrieving static content the app you are using to retrieve it will receive the query strings found in a link? or they should be ignored and only the content returned to the client app? 
E.g., let's say I have `safe://<xorurl1>/index,html?v=3` which shows who won a lottery based on a query string `winner=<name>`. Then let's say I publish the winner at `safe://thewinner` by linking that NRS name to `safe://xorurl1>/index.html?v=3&winner=gabriel`. What do we want to happen when I enter `safe://thewinner` in the browser? shall it receive not only the content of the `index.html?v=3` or also the query string that was in the NRS name? 

I think the answer can only be:
a- we always ignore the query string in a NRS link
b- we always respect it and pass it back to the app fetching the target content, regardless if the url contained the same query string.

**Comment by dan-da** [comment]
So, yeah, I think we should start a discussion in the forum to get more input.    But I will make a final comment here for the record before doing that.  I will go along with what the community thinks best after discussion.

So the example above is a tad bit contrived, I think.  The winner's name should be stored inside index.html (or other suitable data type), not in a query string.  Missing from the example is that somewhere there is an app that will interpret the URL, as per rfc-0053, to actually show the winner. The first URL doesn't magically do it by itself.  But regardless, there is still a way that scheme could be made to work with the ruleset I proposed:

1. The fully resolved url the app would see for http://thewinner is: safe://<xorurl>/index.html?v=3&winner=gabriel.     ( as you wish )
2. The fully resolved url the app would see for http://thewinner?winner=bochaco is safe://<xorurl>/index.html?v=3&winner=bochaco.  ( which you don't like, I believe )
3. An app that only cares about the first time a query-string param is defined, could call Safe::inspect() on the url to determine that.  Which would give safe://<xorurl>/index.html?v=3&winner=gabriel

My argument is that devs and users alike are used to being able to provide query params in the browser, and have those take effect. eg, typically url params override form POST params, for example, or defaults embedded in web-apps.   Using your ruleset would sort of reverse or break this model, effectively taking power away from the end user.

Regarding app usage of the url, I found these comments in [rfc-0053](https://github.com/maidsafe/rfcs/blob/master/text/0053-xor-urls/0053-xor-urls.md), which maybe shed a bit more light, if not exactly taking a side in this present debate:

> Since the main goal is to have a standard way of referencing and linking any piece of data stored on the SAFE Network with a URL (as required for supporting linked-data), thus the <query> and <fragment> parts are reserved to be used by the application consuming the content rather than used by the resolver itself. Once the resolver has located the content and fetched it, the values provided in the <query> and <fragment> parts shall be passed back to the application which requested the content. The application can then define what it's the way of referring to a specific part of the document you are targeting with the URL. E.g. the WebID spec makes use of the #fragment to refer to one of the subgraphs, i.e. one of the persons/agents described in the WebID Profile document. As another example, the #fragment and query values are used by the web page that is being referenced by a URL and not used by the DNS to locate the file/s of the web page. In the case of the SAFE Network, the <fragment> could be used to refer to a specific entry of a MutableData, or the <query> can also be used to provide a filter to be applied to the entries that need to be considered on a referenced MutableData. This all would be defined/determined by the convention followed by the application that is fetching the content from a XOR-URL.

and 

> #somesection?somekey=5 is the <fragment> and <query> parts which will be ignored by the resoler and simply returned back to the application for it to use it.

**Comment by joshuef** [comment]
> query string should always precede fragment in a url, so I think that sentence is in error, but whatever. :-)

That can be happily updated in the RFC.

To keep things simple here... why does the resolver need the query param at all (beyond version... which is a special case which we'd wanted to put elsewhere [in the port position of the URL was the dream, but that's limited to 65553 or some such]....)

Anywho, query params are send to the server in clearnet, so the server can respond with different data. We do not have this capability (beyond the version, as above, I'd choose to ignore this for the general convo as we could still decide to put that _elsewhere_ if we wanted to...). 

So my gut is just to ban any query param from being added to an NRS url. I don't think we loose anything by doing that aside from complexity. 

They can still be applied in browser. They can still be reacted to by a webapp. NRS resolution is simpler.

----

If we want to consider query params as they are used today, then `?v=` is _the only one_ we allow. Cos the rest don't do what they do today (ie change content served)

**Comment by dan-da** [comment]
> To keep things simple here... why does the resolver need the query param at all

Because, given status quo, it is resolving links that people have already stored, possibly with query params.

Now, as you point out, one possibility is to prohibit storing them at all, and indeed it may be simplest to do.   I think though, that this is less flexible and may prove short-sighted, as we cannot know all the possible application use-cases people will build, or want to build.   And I see a lot of value in being able to use an NRS url as a "bookmark" to accessing a particular app in a particular way.   For example, think of an app like google maps that takes ?lat and ?long and ?view query params.

So we can create an nrs bookmark for a particular map, eg:
`safe://treasuremap   --> safe://xor.../mapdata?lat=1.234&long=-0.2344&view=satellite`

Ok, so let's compare/contrast:
* with @joshuef's rules, such a bookmark link could not be created, only treasuremap --> mapdata?v=0.    End user must enter safe://treasuremap?lat=1.234&long=-0.2344&view=satellite.  So the utility/value of the "bookmark" is diminished.
* with @bochaco's rules, this bookmark could be created, but end-user (or a subsequent link) has no way to override any query param.  the query string in bookmark is final.
* with my rules, this bookmark can be created, and end-user (or a subsequent link) can override any query param, eg I can use:   safe://treasuremap?view=streets

**Comment by joshuef** [comment]
Some thoughts:

> Because, given status quo (as defined in rfc-0053 and implemented in safe nrs create), it is resolving links that people have already stored, possibly with query params.

I don't think status quo is necessarily a good reason to choose to do something if we're seeing it being problematic. All data will be wiped as we move forward with testnets. Folk are aware this is alpha stuff. If the RFC says one thing, but we think that's wrong, there's no issue with updating it, or superseding it IMO.

----

> as we cannot know all the possible application use-cases people will build, or want to build

True. 

But nature finds a way. 🦕 

There have already been mentioned other possible ways to achieve `safe://treasuremap` defaulting to certain params. (A new html page..., a 'bookmarks' page linking _with_ params)

What other use cases do we want to cater for and why?

----

I'd honestly be in favour of banning query-params from NRS _just_ to avoid any server-side confusion... This idea that query params can change what content is served to you (as on clearnet), when it does not do that at all.

Though that'd still leave `#fragments` which can be used to the same effect, so it doesn't really solve any issues, but perhaps _clarifies_ one thing.

---

> My argument is that devs and users alike are used to being able to provide query params in the browser, and have those take effect. 

There's nothing stopping this without query params in NRS resolution. Just the app has to handle it.

----

If you `safe cat safe://treasuremap` (with the NRS _having_ the `fragment`) you dont get back any different data than without it... which will be different to browser behaviour if the `fragment` from NRS maps is needing to be default applied there.

So in the end, dealing with `param/fragment` is a App layer concern IMO. Not strictly something for NRS to be handling (as it cannot).

----

All of the above is not to say that I don't find the 'default args applied' case to be interesting....

If we actually want something like `default fragment` in our NRS, perhaps it should be separate to the URL resolution? (And therefore more clearly separated into the app layer)? There could be a `defaultFragment` entry in the NRS map RDF for a given URL.

(I'm aware that doesn't answer the question or which to be applying / what overrides what, but the separation may be helpful in terms of reasoning about this).

**Comment by bochaco** [comment]
> So in the end, dealing with param/fragment is a App layer concern IMO. Not strictly something for NRS to be handling (as it cannot).

I think I'm buying this, this is the key, I think we are loosing focus on what NRS is, subnames and path is already enough for NRS, the rest can be handled by apps as they whish, so safe://myname?k1=v1 would be resolved exactly as safe://myname and then k1=v1 is the app's responsibility to handle it or not.

**Comment by dan-da** [comment]
> There have already been mentioned other possible ways to achieve safe://treasuremap defaulting to certain params. (A new html page..., a 'bookmarks' page linking with params)

**Comment by dan-da** [comment]
> There have already been mentioned other possible ways to achieve safe://treasuremap defaulting to certain params. (A new html page..., a 'bookmarks' page linking with params)

true, but why add/require another level of indirection for this common/popular use case?

> What other use cases do we want to cater for and why?

It seems that NRS url is presently catering to a present need of "?v" for bookmarking specific version.  As far as I can tell this was the main/only reason to include query string to begin with, no?  And who's to say we don't come up with other special case bookmark needs in the future?

> I'd honestly be in favour of banning query-params from NRS just to avoid any server-side confusion...

Ok, then please propose a simple set of rules covering each remaining nrs url component, that we can evaluate and reason against, as I did in the OP.  For example 'path' and '?v'.  Presently, I feel you both are arguing against what I proposed, without creating a cohesive counter-proposal, so I am debating strawmen.

> This idea that query params can change what content is served to you (as on clearnet), when it does not do that at all.

I'm not sure I fully agree with this statement.   First of all, on clearnet, query params do not change returned content of a static html file, for example.  It is only when the endpoint is a web-app of some kind that query params take effect.  I see a similar (not identical) distinction with SAFE network.  The content returned from network is static (already with an exception of ?v), but the app that a user is interacting with may choose to use the query params to modify the displayed content in interesting ways.  Most end users will not really know or care that it is a local app interpreting the param rather than a remote app.  

Ok the above is all kind of fun to debate about, but I don't feel it really gets down to the core of this issue yet though, which is:  what is an NRS url, and what is its purpose?

NRS: Name Resolution System.  right?   sounds like a replacement for DNS.  But if it were only that, then why use a URL form at all?   We should just be mapping name --> xor (sans scheme, path, query, etc).   @joshuef I think you are arguing in this direction at least, if not taken to logical conclusion yet.   So anyway, I view that as a decent option I could probably support.  Pretty simple and clean.  Very close to DNS (name --> IP) model.

But what we presently have is rather url (with name)  --> url ( with name or xor ).  And URLs are composed of scheme, auth, host, path, query, fragment.    Why was this form chosen instead of the simpler DNS-like model above?  I figured it was for the bookmark type use-case.  If not, maybe you guys can explain the reasoning to me.  But what I do see is that it is inconsistent and arbitrary to have a url(a) ---> url(b) model, but then require that url(a) use "?v" in a special way, not define/allow any other query params, and concat a's path to b's path.

In other words, I feel like if we are going with a url --> url model, we should fully embrace and support it, using something like my OP ruleset.  Otherwise we end up with weirdnesses like paths are supported but not query strings.  Failing that, a name --> xor model seems better to me.  or name --> (name or xor) for link chains.

So to be as clear as possible, this ruleset would simply be:

```
1. name  --> (name or xor)   (not a url, just the xorname component)
2. name can have sub-names, eg bob.hr.maidsafe that links to its own (name or xor).
3. name can be used within 'host' portion of a url only.
```
A fallout from the above is that ?v is no longer a special case bookmark, and would need to be explicitly requested, for better or worse.

**Comment by joshuef** [comment]
> I'm not sure I fully agree with this statement. First of all, on clearnet, query params do not change returned content of a static html file, for example. It is only when the endpoint is a web-app of some kind that query params take effect

Sure, apps don't have to use params in that way, but that is the _intention_ of params (query params are sent to server to aid processing and returning of a response; hash fragments are not). Though hash fragments, which can be used in the same way as you're proposing for query params (in the web-app utilising them). 

My point was that we have _other_ means of doing using URL as a ui for webapps to react to with hash fragments. That we do not need query params and the baggage of their (potential) server implications (for web devs if nothing else). 

Though that may be a draconian suggestion. 

----
> Ok, then please propose a simple set of rules covering each remaining nrs url component

In general I'd proporse to stay with NRS as described here: https://github.com/maidsafe/rfcs/blob/master/text/0052-RDF-for-public-name-resolution/0052-RDF-for-public-name-resolution.md#url-resolution (which does not talk about query params for NRS resolution). But it does cede to 'xorurl' resolution in the first instance.

Actually just [re reading the xor url rfc](https://github.com/maidsafe/rfcs/blob/master/text/0053-xor-urls/0053-xor-urls.md) :

> <query>: query arguments, to be used by the client app and not for retrieving the content

Which is contrast to another part quoted by @dan-da above

> Once the resolver has located the content and fetched it, the values provided in the <query> and <fragment> parts shall be passed back to the application which requested the content.

Which is basically saying the opposite (or so it reads to me at least).

> But if it were only that, then why use a URL form at all? We should just be mapping name --> xor (sans scheme, path, query, etc)

Hmmmm. Yeh, i see what you're getting at here. NRS itself is just that, I would say. A mapping to a url from a name. (Perhaps the fact it's put in the context of `safe://` confuses things as those names _become implicitly_ urls...).

For names your proposal 

```
1. name  --> (name or xor)   (not a url, just the xorname component)
2. name can have sub-names, eg bob.hr.maidsafe that links to its own (name or xor).
3. name can be used within 'host' portion of a url only.

```
Seems good to me.

But then you have your _version_ that needs to be passed on.

A slight modification of the original set proposed by @dan-da (where I'll say 'version' instead of v query param; as it may come out that we need to change that for clarity?)

    - The path of a nrs-url is appended to the path of linked-url
    - The version of nrs-url is used to lookup _the linked FilesContainer.
    - The version is _not_ applied to any of the links from that `FilesContainer`. They all retain their links (which they must have on anything non ImmutableData to be stored in a valid `FilesContainer`). The versions listed in the chosen `FilesContainer` (at that requested version) are respected.
    - Query params and hash fragments have no influence over resolution of URLs, these are app level constructs.

**Comment by dan-da** [comment]
One point of clarification:  In my last comment I proposed mapping a name to another name or an xor *identifier*, NOT to a url.  I think that mapping to a URL is what gets us into weird situations.

In other words, I think the mapping should take place only within the 'host' portion of a URL.  All the rest of the URL is unaffected.    Just as DNS maps name --> IP, and that's it.

Else, if we map a name to a URL, it is taking the 'host' field of one URL and putting an entire URL in its place.   That's just plain weird in my book.

**Comment by bochaco** [comment]
I think by doing that you'll be breaking the possibility of having linked-data and RDF-like data formats, XOR-URLs are the key for achieving that. We used to have an NRS before which did name -> xorname and that didn't allow us to have standard RDF/linked-data formats, I think that's the main reason for having URLs anywhere you are linking one piece of SAFE content with another.

**Comment by dan-da** [comment]
rfc-0052 was [updated](https://github.com/maidsafe/rfcs/pull/356) with further clarifications/examples, and also the code in #535.   closing.
